Sources/GRPC/ConnectionPool/ConnectionPool.swift

```diff
@@ -502,15 +509,30 @@ internal final class ConnectionPool {
       self._state = .shuttingDown(promise.futureResult)
       promise.futureResult.whenComplete { _ in
         self._state = .shutdown
+        self.delegate = nil
         self.logger.trace("finished shutting down connection pool")
       }
 
       // Shutdown all the connections and remove them from the pool.
       let connections = self._connections
       self._connections.removeAll()
 
-      let allShutdown = connections.values.map {
-        $0.manager.shutdown(mode: mode)
+      let allShutdown: [EventLoopFuture<Void>] = connections.values.map {
+        let id = $0.manager.id
+        let manager = $0.manager
+
+        return manager.eventLoop.flatSubmit {
+          // If the connection was idle/shutdown before calling shutdown then we shouldn't tell
+          // the delegate the connection closed (because it either never connected or was already
+          // informed about this).
+          let connectionIsInactive = manager.sync.isIdle || manager.sync.isShutdown
```

Comment:
Is this check too early? Is it possible for us to quiesce between calling `shutdown` and the future completing?

Reply:
Good question, are you getting at the possibility of registering callbacks to the delegate multiple times?

- Call `shutdown()`
- Hit this line, register callbacks to the delegate
- Server tells us to quiesce before future completes
- Code handling quiescing registers callbacks to the delegate

If so that shouldn't happen: we drop all connection state a few lines above (L510/L518) and the quiescing code exits early if the connection state isn't present.